ops/manifests/manifest.py

```diff
@@ -0,0 +1,329 @@
+        for manipulate in self.manipulations:
+            if isinstance(manipulate, Addition):
+                obj = manipulate()
+                if not obj:
+                    continue
+                result[HashableResource(obj)] = None
+
+        # From static manifests
+        for manifest in (self.manifest_path / ver).glob("*.yaml"):
+            for obj in self._safe_load(manifest):
+                result[HashableResource(obj)] = None
+
+        return result.keys()
+
+    @lru_cache()
+    def _safe_load(self, filepath: Path) -> List[AnyResource]:
+        """Read manifest file and parse its content into lightkube objects.
+
+        Lightkube can't properly read manifest files which contain List kinds.
+        """
+        content = filepath.read_text()
+        return [
+            codecs.from_dict(item)  # Map to lightkube resources
+            for rsc in yaml.safe_load_all(content)  # load content from file
+            if rsc  # ignore empty objects
+            for item in (rsc["items"] if rsc["kind"] == "List" else [rsc])
+        ]
+
+    def status(self) -> FrozenSet[HashableResource]:
+        """Returns all installed objects which have a `.status.conditions` attribute."""
+        return frozenset(_ for _ in self.installed_resources() if _.status_conditions)
+
+    def installed_resources(self) -> FrozenSet[HashableResource]:
+        """All currently installed resources expected by this manifest."""
+        result: Dict[HashableResource, None] = OrderedDict()
+        for obj in self.resources:
+            try:
+                next_rsc = self.client.get(
+                    type(obj.resource),
+                    obj.name,
+                    namespace=obj.namespace,
+                )
+            except ApiError:
+                continue
+            result[HashableResource(next_rsc)] = None
+        return frozenset(result.keys())
+
+    def labelled_resources(self) -> FrozenSet[HashableResource]:
+        """Any resource ever installed and labeled by this class."""
+
+        NamespaceKind = namedtuple("NamespaceKind", "namespace, kind")
+        ns_kinds = set(
+            NamespaceKind(obj.namespace, type(obj.resource)) for obj in self.resources
+        )
+
+        return frozenset(
+            HashableResource(rsc)
+            for ns_kind in ns_kinds
+            for rsc in self.client.list(
+                ns_kind.kind,
+                namespace=ns_kind.namespace,
+                labels={
+                    "juju.io/application": self.app_name,
+                    "juju.io/manifest": self.name,
+                },
+            )
+        )
+
+    def apply_manifests(self):
+        """Apply all manifest files from the current release after manipulating."""
+        log.info(f"Applying {self.name} version: {self.current_release}")
+        for rsc in self.resources:
+            for manipulate in self.manipulations:
+                if isinstance(manipulate, Patch):
```

Comment:
why patch here and not in resources?

Reply:
hmm.  yes, i suppose that's reasoned...  i don't have a reason "why". 

I'll think deep thoughts on this

Reply:
Done!